### ISSUE TITLE
MacOS version does not open files when double clicked

### ISSUE THREAD

**Issue by rhys073** [helge17/tuxguitar#99]
All files opens untitled.tg in Tuxguitar

When tuxguitar.sh is called from the Terminal with a relevant argument, TuxGuitar will open the file, however double-clicking a file (calling the .app) presumably does not pass through the /path/filename arguement to java

**Comment by guiv42** [comment]
I can reproduce something similar on Linux with 1.6.0beta1.
bisect leads me to 2cfb9c919119eedece46eb52d656c88000b4af66, but to be honest I do not really understand. Investigation in progress...

**Comment by guiv42** [comment]
Could anyone please test [this version](https://github.com/guiv42/tuxguitar/tree/fix99) in MacOS environment? (I can't)
It seems to solve the issue in Linux environment.

Issue seems critical, so I just reverted 2cfb9c919119eedece46eb52d656c88000b4af66. I can take some time later to try to reintroduce that feature.
Not fully understood yet, but it seems it introduced a deadlock or some similar issue in the init process (not clear why it only appears when launching TG by double-clicking a file)

**Comment by helge17** [comment]
The two issues are not related, imho.

### TuxGuitar on MacOS not opening the file selected in Finder

As far as I understand, the file associations and the startup script are defined in the Contents/Info.plist file:
```
<key>CFBundleExecutable</key>
<string>tuxguitar.sh</string>
```

tuxguitar.sh reads the file to open from "$@", the command arguments. Now if you add something like

```
echo "# TEST"         > /tmp/test
echo "First arg: $1" >> /tmp/test
echo "Alls args: $@" >> /tmp/test
```

to tuxguitar.sh and click on a TuxGuitar file in Finder, you can see in /tmp/test, that there are no arguments: Finder just doesn't pass any arguments to tuxguitar.sh, so TuxGuitar has no way of knowing that it should open a file. At the moment I have no idea where TuxGuitar can get this information from or how to start TuxGuitar properly.

### TuxGuitar on Linux/Windows/MacOS/BSD hanging when starting with a file

I have tested with two files containing only one measure; one without marker, the other with one single marker. The problem does not seem to occur without markers, but also occurs quite rarely (10% of the time?) with the file with only one marker.

I have also tested the behavior with other files containing a complete song, and indeed it seems to be related to the markers: the more markers there are in the file, the more often TuxGuitar hangs.

When opening files containing many markes with 1.6.0beta1 (SWT, deb package), TuxGuitar hangs nearly on every start and shows the following error:

```
java.lang.IllegalArgumentException: Argument not valid
        at org.eclipse.swt.SWT.error(SWT.java:4692)
        at org.eclipse.swt.SWT.error(SWT.java:4626)
        at org.eclipse.swt.SWT.error(SWT.java:4597)
        at org.eclipse.swt.graphics.GC.drawImage(GC.java:838)
        at org.herac.tuxguitar.ui.swt.resource.SWTPainter.drawImage(SWTPainter.java:90)
        at org.herac.tuxguitar.app.view.component.table.TGTableHeaderMeasures.paintControl(TGTableHeaderMeasures.java:110)
        at org.herac.tuxguitar.app.view.util.TGBufferedPainterLocked.fillPaintBuffer(TGBufferedPainterLocked.java:50)
        at org.herac.tuxguitar.app.view.util.TGBufferedPainterLocked.paintBufferLocked(TGBufferedPainterLocked.java:59)
        at org.herac.tuxguitar.app.view.util.TGBufferedPainterListenerLocked.onPaint(TGBufferedPainterListenerLocked.java:14)
        at org.herac.tuxguitar.ui.event.UIPaintListenerManager.onPaint(UIPaintListenerManager.java:33)
        at org.herac.tuxguitar.ui.swt.event.SWTPaintListenerManager.paintControl(SWTPaintListenerManager.java:20)
        at org.eclipse.swt.widgets.TypedListener.handleEvent(TypedListener.java:234)
        at org.eclipse.swt.widgets.EventTable.sendEvent(EventTable.java:89)
        at org.eclipse.swt.widgets.Display.sendEvent(Display.java:5618)
        at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1405)
        at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1431)
        at org.eclipse.swt.widgets.Widget.sendEvent(Widget.java:1414)
        at org.eclipse.swt.widgets.Control.gtk_draw(Control.java:3940)
        at org.eclipse.swt.widgets.Scrollable.gtk_draw(Scrollable.java:342)
        at org.eclipse.swt.widgets.Composite.gtk_draw(Composite.java:455)
        at org.eclipse.swt.widgets.Widget.windowProc(Widget.java:2217)
        at org.eclipse.swt.widgets.Control.windowProc(Control.java:6846)
        at org.eclipse.swt.widgets.Display.windowProc(Display.java:5842)
        at org.eclipse.swt.internal.gtk.GTK._gtk_main_do_event(Native Method)
        at org.eclipse.swt.internal.gtk.GTK.gtk_main_do_event(GTK.java:4154)
        at org.eclipse.swt.widgets.Display.eventProc(Display.java:1474)
        at org.eclipse.swt.internal.gtk.OS._g_main_context_iteration(Native Method)
        at org.eclipse.swt.internal.gtk.OS.g_main_context_iteration(OS.java:1603)
        at org.eclipse.swt.widgets.Display.readAndDispatch(Display.java:4404)
        at org.herac.tuxguitar.ui.swt.SWTApplication.start(SWTApplication.java:64)
        at org.herac.tuxguitar.app.TuxGuitar.createUIContext(TuxGuitar.java:116)
        at org.herac.tuxguitar.app.TuxGuitar.createApplication(TuxGuitar.java:101)
        at org.herac.tuxguitar.app.TGMainSingleton.launchTuxGuitar(TGMainSingleton.java:57)
        at org.herac.tuxguitar.app.TGMainSingleton.launchSingleton(TGMainSingleton.java:45)
        at org.herac.tuxguitar.app.TGMainSingleton.main(TGMainSingleton.java:35)
```

The behavior is the same if you start TuxGuitar from the command line (tuxguitar song_with_markers.tg) or by clicking on song_with_markers.tg in the file manager (in my case KDE Dolphin).

The error does not occur if TuxGuitar is already running. Then, as expected, a new tab is opened containing the given file.

The error occurs on Linux, FreeBSD, MacOS and Windows.

The error occurs only with SWT, not with JFX.

@guiv42, I also tested your fix99 version: The "hang" error no longer occurs, the problem that TuxGuitar does not open the selected file on MacOS is still there.

**Comment by guiv42** [comment]
thank you @helge17 for this detailed analysis, that's very clear.
About Linux and markers, I will submit a new pull request to fix the issue, still keeping the markers displayed (it's currently under test)

Regarding MacOS issue: do you know if it was correctly working with 1.5.6? (I don't have in mind any bug report of this kind in SF)

**Comment by helge17** [comment]
Same problem with 1.5.6: When clicking on a TG file in Finder, TuxGuitar starts with an empty file.

**Comment by guiv42** [comment]
"**TuxGuitar on Linux/Windows/MacOS/BSD hanging when starting with a file**" issue should be solved by #100.
Issue seems quite critical to me, so it could deserve a beta2 release.  @helge17: it's up to you, I don't know how complex and time-consuming the full delivery process is

About the MacOS issue unfortunately I don't think I can help in any way. But at least it's not a bug we have introduced, and since it's been there for a while it may be treated with lower priority I think.

**Comment by helge17** [comment]
Thanks for the patch; release 1.6.0beta2 is now online.

**Comment by helge17** [comment]
I have tried two approaches to fix the "TuxGuitar on MacOS not opening the file selected in Finder" issue, but neither are ready for commit.

1. Use an AppleScript wrapper to pass filenames to tuxguitar.sh

Basically this worked, you can double click on a TG/GP file in Finder (or drop it on the TuxGuitar app) and TuxGuitar will open the file. But two small problems remain:

- When TuxGuitar is launched, a shell icon appears in the Dock for a brief moment (~half a second). After that the correct TuxGuitar icon is displayed.
- After starting TuxGuitar, the menu bar is blocked and does not respond to clicks. The problem disappears when you click into another application and then back to TuxGuitar: After that, the menu behaves normally. This seems to be a known issue (https://stackoverflow.com/questions/19803332/menu-doesnt-show-up-with-swt-on-mac-os-x) and may be fixed in future SWT versions, if I understand correctly (https://github.com/eclipse-platform/eclipse.platform.swt/issues/256).

You can reproduce the second problem by running tuxguitar.sh from the MacOS terminal app.

2. Create a pkg package

This would be the modern way to distribute applications for MacOS anyway. I was able to create a simple pkg using jpackage. But that doesn't solve our initial problem; it seems that you also have to modify the Java code so that TuxGuitar can take over the filenames from Finder:

https://stackoverflow.com/questions/1575190/double-click-document-file-in-mac-os-x-to-open-java-application

So for version 1.6.x there probably won't be a fix, but for further versions there is hope ;-)

**Comment by guiv42** [comment]
I also searched a bit the Internet on this topic. I have zero knowledge of Apple's ecosystem, but it seems the solution is not in a command line parameter but an Apple event.
It seems current code defines [dedicated classes](https://github.com/helge17/tuxguitar/tree/master/TuxGuitar-cocoa-integration/src/org/herac/tuxguitar/cocoa) to manage this, even an "opendoc" folder...

Could you please check what happens when double-clicking a file whilst TG is already opened?
If this file opens correctly it may be a timing issue between the app initialization and when the event is sent by MacOS

**Comment by helge17** [comment]
It makes no difference whether TuxGuitar is running or not, the file will not be opened by Finder (the MacOS file manager) in either case.

tuxguitar.sh expects the file to be opened as an argument and then passes the filename to Java. Finder does not pass the filenames via arguments, but by other means.

You can open a file from the command line with:

`% open /path/to/tuxguitar-1.6.0beta3-macosx-swt-cocoa-x86_64.app --args /path/to/tab.tg`

or

`% /path/to/tuxguitar-1.6.0beta3-macosx-swt-cocoa-x86_64.app/Contents/MacOS/tuxguitar.sh /path/to/tab.tg`

I think for a clean solution we need to get rid of the shell script and start Java directly "the Apple way". For example jpackage could do the job. But I don't know the Apple ecosystem, nor can I really code, so I could well be wrong ;-)

By the way, if you put more than one file on TuxGuitar, only the first file will be opened, not all of them, on all operating systems.

**Comment by G-eos** [comment]
I try do dig a little on this subject. I have no background with java application development, nor with Apple development not tux guitar. Thus this can be completely out of meaning.
To go quick, I ask ChatGPT if he has some good hints. As expressed above, it seems that MacOS use a special event for this purpose as @guiv42 you highlight before.

Please find ChatGPT hints:

When an application is launched by double-clicking a file associated with it in Finder on macOS, the file's path is not directly provided as a command-line argument to the application. Instead, macOS launches the application and sends an Apple Event to it containing information about the file that was double-clicked.

To handle file opening events like this in a Java application on macOS, you typically need to use the Apple-specific Java Application Events API. This API allows your Java application to receive and handle Apple Events, including those related to file opening.

Here's a basic example of how you might use the Java Application Events API to handle file opening events in a Java application on macOS:

```
import com.apple.eawt.AppEvent;
import com.apple.eawt.AppEventListener;
import com.apple.eawt.Application;

public class MyApp {
    public static void main(String[] args) {
        // Register an application event listener
        Application.getApplication().setAppEventListener(new AppEventListener() {
            @Override
            public void handleOpenFile(AppEvent.OpenFilesEvent event) {
                // Handle file opening events
                for (java.io.File file : event.getFiles()) {
                    String filePath = file.getAbsolutePath();
                    System.out.println("File opened: " + filePath);
                    // Process the file as needed
                }
            }
        });
    }
}
```
In this example, we're using classes from the com.apple.eawt package, which is part of the Apple Java Extensions. These classes provide access to macOS-specific functionality, including handling file opening events.

When a file is opened by double-clicking on it in Finder, the handleOpenFile method in the AppEventListener is called, and you can access the file's path from the event.getFiles() method. You can then process the file as needed within your application.

It's worth noting that the com.apple.eawt package is specific to macOS and may not be available on other platforms. Therefore, you may need to use platform-specific code or libraries if you're targeting multiple operating systems.

**Comment by G-eos** [comment]
> I also searched a bit the Internet on this topic. I have zero knowledge of Apple's ecosystem, but it seems the solution is not in a command line parameter but an Apple event. It seems current code defines [dedicated classes](https://github.com/helge17/tuxguitar/tree/master/TuxGuitar-cocoa-integration/src/org/herac/tuxguitar/cocoa) to manage this, even an "opendoc" folder...
> 
> Could you please check what happens when double-clicking a file whilst TG is already opened? If this file opens correctly it may be a timing issue between the app initialization and when the event is sent by MacOS

With Tuxguitar 1.6.1:
- Double click without TuxGuitar launched: launch it with a new.
- Double click with TuxGuitar launched: Does "nothing", just place TuxGuitar in front of the Windows Manager and give focus

**Comment by G-eos** [comment]
> To handle file opening events like this in a Java application on macOS, you typically need to use the Apple-specific Java Application Events API. This API allows your Java application to receive and handle Apple Events, including those related to file opening.
> 
east has been deprecated ... see there https://openjdk.org/jeps/272
need to to use awt ...

**Comment by G-eos** [comment]
It seems that https://github.com/AvaloniaUI/Avalonia/discussions/5504#discussioncomment-1552446 provides some interesting information.
We need to register to Application.UrlsOpened event and this has to be done as soon as possible in the startup of the Application to not miss the event.

Now, how to register to an event, and where to add this code :)

**Comment by guiv42** [comment]
> this has to be done as soon as possible in the startup of the Application to not miss the event

App's entry point is [there](https://github.com/helge17/tuxguitar/blob/8c5f1200d549b05f34ce335e06650e6831f6cfbb/desktop/TuxGuitar/src/org/herac/tuxguitar/app/TGMainSingleton.java#L27), however I have no idea where the first call to OS-specific code can be found...

**Comment by G-eos** [comment]
After some digging, and following obsolete API. The good one since Java 9 since ok in Java 20 seem to be:
[java.awt.desktop.OpenFilesHandler](https://docs.oracle.com/en/java/javase/20/docs/api/java.desktop/java/awt/Desktop.html#setOpenFileHandler)

Usage seems to be done [here](https://github.com/cryptomator/cryptomator/blob/b40e6db70179dd06a47ed8f7c9b7c6494684cdd4/src/main/java/org/cryptomator/launcher/FileOpenRequestHandler.java#L38) or [There](https://stackoverflow.com/questions/77646000/open-a-file-with-java-application-on-macos/77646772#77646772)

I try to add into desktop/TuxGuitar-cocoa-integration-swt/src/org/herac/tuxguitar/cocoa/opendoc/OpenDocListener.java the following code without great success. It compiles, run, I see the Test INIT on console but never see something when try to drag tuxguitar files on the opened app. When double-click it launch another TuxGuitar which stop immediately :-(

```
import java.awt.Desktop;
import java.awt.desktop.OpenFilesEvent;
import java.awt.desktop.OpenFilesHandler;
...
public void OpenMyAppFilesHandler(OpenFilesEvent e) {
		System.out.println("OpenFileHandler()");
	 }
...
	public void init() throws Throwable{
		Desktop.getDesktop().setOpenFileHandler(this::OpenMyAppFilesHandler);
		System.out.println("Test INIT OK");
....
```

Does someone can bring some help :)

**Comment by guiv42** [comment]
Could you please push your modifications in a dedicated (experimental) branch, so I can have a look?
(knowing I won't be able to build nor to test...)

**Comment by G-eos** [comment]
I will, but this is absolutely ugly :)

**Comment by guiv42** [comment]
Just reading the java doc you have mentioned:
`public void setOpenFileHandler(OpenFilesHandler openFileHandler)`

I understand the argument you need to pass is an instance of a class which implements [OpenFilesHandler](https://docs.oracle.com/en/java/javase/20/docs/api/java.desktop/java/awt/desktop/OpenFilesHandler.html) interface.

This could look like this
warning: NOT built, tested or whatever, not even written in an IDE checking syntax, I can also do ugly things ;)
```
private class MyOpenFilesHandler implements OpenFilesHandler {   // should require one more "import" statement?
    void openFiles(OpenFilesEvent e) {
        // here, try to print some attribute of e, to check you get the expected info
        // see https://docs.oracle.com/en/java/javase/20/docs/api/java.desktop/java/awt/desktop/OpenFilesEvent.html
    }
}

(...)
	public void init() throws Throwable{
		Desktop.getDesktop().setOpenFileHandler(new MyOpenFilesHandler());
		System.out.println("Test INIT OK");
```

note: a short test like `getDesktop().isDesktopSupported()` could prove useful (?)

**Comment by G-eos** [comment]
It's there https://github.com/G-eos/tuxguitar/tree/wip/mac_open_files

**Comment by G-eos** [comment]
I try your code, does not compile. But I am not a java coder, probably trivial ... the code is pushed on the branch

[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.10.1:compile (default-compile) on project tuxguitar-cocoa-integration-swt: Compilation failure: Compilation failure: 
[ERROR] /Users/xav/git/tuxguitar/desktop/TuxGuitar-cocoa-integration-swt/src/org/herac/tuxguitar/cocoa/opendoc/OpenDocListener.java:[9,9] modifier private not allowed here
[ERROR] /Users/xav/git/tuxguitar/desktop/TuxGuitar-cocoa-integration-swt/src/org/herac/tuxguitar/cocoa/opendoc/OpenDocListener.java:[10,10] openFiles(java.awt.desktop.OpenFilesEvent) in org.herac.tuxguitar.cocoa.opendoc.MyOpenFilesHandler cannot implement openFiles(java.awt.desktop.OpenFilesEvent) in java.awt.desktop.OpenFilesHandler
[ERROR]   attempting to assign weaker access privileges; was public

**Comment by guiv42** [comment]
replace
`void openFiles(OpenFilesEvent e) {`
by
`public void openFiles(OpenFilesEvent e) {` ?

**Comment by G-eos** [comment]
Thanks !
I will have a look tomorrow

**Comment by guiv42** [comment]
https://github.com/G-eos/tuxguitar/pull/1

**Comment by G-eos** [comment]
Some progress here.  Look at the branch [mac_open_files](https://github.com/G-eos/tuxguitar/tree/wip/mac_open_files). Latest [commit](https://github.com/G-eos/tuxguitar/commit/8371ee3dcedeb37ff47c8878a818b65f14c1cd52).
Double click on a .tg file launch TuxGuitar without catching the event.
But all later double click lead to an event triggered :-)
The log files is

```
addListener OpenDocument
OMG Event !!!!!!
Event {type=46 null time=428746473 data=null x=0 y=0 width=0 height=0 detail=0}
```
Note: anoure files -> only the time change. 

Good doc for this event is [HERE](https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Fguide%2Fproduct_open_file.htm).

This sounds promising 

Note: No way to catch event with AWT, SWT take the lead. We shall stay with it.

**Comment by guiv42** [comment]
OK, good news!
> Double click on a .tg file launch TuxGuitar without catching the event.

Possibly a timing issue. I'll have a look at your branch, and see if we can trigger the registration of a listener earlier in the init process.
Just one question: why did you register the "openDocument" listener in MacMenu? Got to check when it is initialized...

**Comment by G-eos** [comment]
> > Double click on a .tg file launch TuxGuitar without catching the event.
> 
> Possibly a timing issue. I'll have a look at your branch, and see if we can trigger the registration of a listener earlier in the init process. And debug log might be a contributor to an init delay... 

In the SWT doc, there explain that the register must be done before "It should register this listener before calling PlatformUI.createAndRunWorkbench ...  before ... starts running the event loop."

> I'll try to propose an alternative implementation based on your work. Just one question: why did you register the 
"openDocument" listener in MacMenu? Got to check when it is initialized...

This was just a test, and reuse the available variables of Menu functions. And as it works, I was very happy to start to catch event and so share ;-)

There is more work needed, first to receive valuable information like the file path.
Todo I see so far:
- [ ] Read again [Eclipse SWT doc](https://help.eclipse.org/latest/index.jsp?topic=%2Forg.eclipse.platform.doc.isv%2Fguide%2Fproduct_open_file.htm)
- [ ] Better parse envents to catch file name and path
- [ ] Connect to TG class to trig Opening the file
- [ ] Move it to OpenDocListener
- [ ] Initialize before "event loop" is started

**Comment by guiv42** [comment]
@G-eos : could you please have a try with my experimental branch?
https://github.com/guiv42/tuxguitar/tree/experimental_mac

Just like before: I still cannot build for mac, so some modifications should be expected

It's inspired from your last experiment. In your todo list, this should check all actions but the last one. This last action is more tricky, as I don't know exactly when the plugin manager initializes the different plugins during app start (and it may be complex to modify...)

**Comment by guiv42** [comment]
> at first open when tuxguitar is not started

not understood. Shall I understand that:
- double-click on a file with TuxGuitar already started => file is opened correctly
- close TuxGuitar, double-click on a file => TuxGuitar opens, but with default template, file is not opened

?

**Comment by G-eos** [comment]
Yes you understand well.

**Comment by guiv42** [comment]
OK, so the most complex use case is still not solved. I've got to look at the initialization of plugins, but I'm not so confident I can change that in a safe manner...

**Comment by G-eos** [comment]
The ability to open a file when TuxGuitar is started is already a good step. Even if it lacks the important part, I agree.

Does this code is specific to Mac ? Does swt does not abstract it, the method are not available if not on Mac or are they empty ?
If yes, it could be done without a plugin ?

No other platforms available to check the idea.

**Comment by guiv42** [comment]
On other platforms, file to open is passed to app as an argument of the launch script. This way of working with events is specific to mac

**Comment by G-eos** [comment]
This is clear. But what happens if this event is registered on the other platform ? Does it crash ? Does not compile ?

**Comment by guiv42** [comment]
> what happens if this event is registered on the other platform ?

I don't know, but just by precaution I would not try to modify something which has been working correctly for years!
It seems OK for me to keep this plugin specific to mac.

I'm currently trying to understand the init sequence. I don't know exactly when the "event loop" is started, but my understanding is that 1st action calling user interface (after splash screen is launched in dedicated thread) is [there](https://github.com/helge17/tuxguitar/blob/ca8bec5cfd865474a3dcc09869c9bc148f8b4d91/desktop/TuxGuitar/src/org/herac/tuxguitar/app/TuxGuitar.java#L135).
Plugins are initialized a few lines below.

One possibility could be to create a category of "early initialization" plugins. In this early init of open document plugin, just catch the event and store it, don't try to trigger any other action. Then, when plugin is "normally" initialized (i.e. as of today), read the stored event and trigger the open action. It seems to me quite close to the example from swt doc you have mentioned above.
Go to have a look at this option (not trivial)

**Comment by G-eos** [comment]
Probably more complex, but clearly safer :-). 
Please note that the splash screen is not always displayed. In particular when TuxGuitar is launched with a double click on a .tg file on the finder.

**Comment by G-eos** [comment]
@guiv42 my last version is your branch, no need to modify. Just compile and run.

**Comment by guiv42** [comment]
Early init of some specific plugins seems feasible, without breaking compatibility with other plugins.
Since implementation would not be trivial, I would like first to make sure the diagnosis is correct. So I implemented a mock-up, where I manually initialize the specific macOS openDoc plugin *before* calling `createWindow()`.

@G-eos : could you have a try with the latest version of this experimental branch? (as usual, not built not tested at all)
https://github.com/guiv42/tuxguitar/commit/4c3676f8481d93728c91d122ccff054050eaad07

I removed all debug logs, in case it does not work it might be needed to re-implement some.

**Comment by G-eos** [comment]
Hello, not better.

I try to move it and end of createMainContext(), it does not work anymore. Even if tuxguitar is already started. It is too early ...

**Comment by guiv42** [comment]
OK thanks for the test. I'm going to set up a test config in Linux, since implementation of openDoc plugin is no more dependent from macOS-specific lib. This should be more efficient...

**Comment by aol-nnov** [comment]
Guys, you could peek the solution from some other java app for mac. For example there is mind mapping app called [freemind](https://sourceforge.net/projects/freemind/) that opens its files from finder. 

It looks somewhat abandoned, but I have v 1.0.1 installed on Sonoma (14.3.1) and it still works like a charm.

**Comment by G-eos** [comment]
> Guys, you could peek the solution from some other java app for mac. For example there is mind mapping app called [freemind](https://sourceforge.net/projects/freemind/) that opens its files from finder.
> 
> It looks somewhat abandoned, but I have v 1.0.1 installed on Sonoma (14.3.1) and it still works like a charm.

Thanks for the hint, the difficulty here is that SWT which provides GFX toolkit seems to have registered the OS event, we must use its own way of doing.
We can have a look at https://github.com/google/gapid/blob/15b44454df3a2469ee9ba42cb116d9b50786c7a4/gapic/src/main/com/google/gapid/util/MacApplication.java#L41 which seems to do the same as what we are trying to do ;-)

**Comment by guiv42** [comment]
@G-eos : I don't understand what configuration your tests refer to, since it seems my last commit does not build.
But nevermind.

Without a context variable, it cannot work. You can try this:
`this.openDocListener = new OpenDocListener(TuxGuitar.getInstance().getContext());`

**Comment by G-eos** [comment]
I tested on the previous commit on your branch.

Thanks a lot for the modification you propose for context. It compiles well.
And, this is a very good news:
- Launch from CLI with a direct call to tuxguitar.sh -> Run without exception, have a stash screen. All further double click on a file leads to set the focus to tuxguitar (menu in menubar is now tuxguitar, but open nothing).
- Launch by double click on a .tg (or gtp, or gp files) leads to launch tuxguitar without splash screen but it open the file :) and an "san titre" empty files -> 2 tabs are opened.
- Lauching tuxguitar apps. And then double click on a file open this file in tuxguitar (and remove the empty one :-)

**Comment by guiv42** [comment]
OK, this is good news indeed!

So, there's still some synchronization issue when opening TG by double-clicking a file: both actions "create new default" and "open double-clicked file" are competing.
It means the opendoc plugin must wait a little more before launching action to open file. Got to find the correct criterion to stop the waiting loop.

**Comment by guiv42** [comment]
@G-eos : could you please try to change the criterion to stop the waiting loop, in OpenDocListener.java:
Replace `while (!pluginManager.allEnabledPluginsInitialized()) {`
by `while (!TuxGuitar.getInstance().isInitialized()) {`

**Comment by guiv42** [comment]
@G-eos : when you have some time, could you please test https://github.com/guiv42/tuxguitar/commit/8da71191a52ca4c4fbac0badd357cc899347ba21?

I tried to cleanup the condition to enable processing of swt events (no more waiting loop). The logic is the following:
- at plugin early init: register swt event listener, and register "action interceptor"
- if, following init, an action to open a file is intercepted, TG was launched with an argument -> stop the plugin (no swt events will ever be received), and let action continue normally.
- if, following init, an action to create an empty file is intercepted, TG was launched without argument. Assume it was launched by Finder (even if not sure). Enable swt events processing when action to create empty file is finished.

If some test scenarios fail, we'll need to add some logging (yet there is none).

**Comment by G-eos** [comment]
At startup from the CLI calling directly tuxguitar.sh plugin seems to crash with this information. I try to debug a bit without great success. Still no Java experience :)

`
java.lang.InstantiationException: org.herac.tuxguitar.cocoa.opendoc.OpenDocPlugin
        at java.base/java.lang.Class.newInstance(Class.java:719)
        at org.herac.tuxguitar.util.TGServiceReader$IteratorImpl.next(TGServiceReader.java:91)
        at org.herac.tuxguitar.util.plugin.TGPluginManager.lookupPlugins(TGPluginManager.java:46)
        at org.herac.tuxguitar.util.plugin.TGPluginManager.<init>(TGPluginManager.java:30)
        at org.herac.tuxguitar.util.plugin.TGPluginManager.<init>(TGPluginManager.java:14)
        at org.herac.tuxguitar.util.plugin.TGPluginManager$1.createInstance(TGPluginManager.java:171)
        at org.herac.tuxguitar.util.plugin.TGPluginManager$1.createInstance(TGPluginManager.java:169)
        at org.herac.tuxguitar.util.singleton.TGSingletonUtil.getInstance(TGSingletonUtil.java:12)
        at org.herac.tuxguitar.util.plugin.TGPluginManager.getInstance(TGPluginManager.java:169)
        at org.herac.tuxguitar.app.TuxGuitar.getPluginManager(TuxGuitar.java:263)
        at org.herac.tuxguitar.app.TuxGuitar.createUIContext(TuxGuitar.java:119)
        at org.herac.tuxguitar.app.TuxGuitar.createApplication(TuxGuitar.java:102)
        at org.herac.tuxguitar.app.TGMainSingleton.launchTuxGuitar(TGMainSingleton.java:57)
        at org.herac.tuxguitar.app.TGMainSingleton.launchSingleton(TGMainSingleton.java:45)
        at org.herac.tuxguitar.app.TGMainSingleton.main(TGMainSingleton.java:35)
Caused by: java.lang.NoSuchMethodException: org.herac.tuxguitar.cocoa.opendoc.OpenDocPlugin.<init>()
        at java.base/java.lang.Class.getConstructor0(Class.java:3761)
        at java.base/java.lang.Class.newInstance(Class.java:706)
        ... 14 more
org.herac.tuxguitar.util.plugin.TGPluginException: An error ocurred when trying to lookup plugin
        at org.herac.tuxguitar.util.plugin.TGPluginManager.lookupPlugins(TGPluginManager.java:51)
        at org.herac.tuxguitar.util.plugin.TGPluginManager.<init>(TGPluginManager.java:30)
        at org.herac.tuxguitar.util.plugin.TGPluginManager.<init>(TGPluginManager.java:14)
        at org.herac.tuxguitar.util.plugin.TGPluginManager$1.createInstance(TGPluginManager.java:171)
        at org.herac.tuxguitar.util.plugin.TGPluginManager$1.createInstance(TGPluginManager.java:169)
        at org.herac.tuxguitar.util.singleton.TGSingletonUtil.getInstance(TGSingletonUtil.java:12)
        at org.herac.tuxguitar.util.plugin.TGPluginManager.getInstance(TGPluginManager.java:169)
        at org.herac.tuxguitar.app.TuxGuitar.getPluginManager(TuxGuitar.java:263)
        at org.herac.tuxguitar.app.TuxGuitar.createUIContext(TuxGuitar.java:119)
        at org.herac.tuxguitar.app.TuxGuitar.createApplication(TuxGuitar.java:102)
        at org.herac.tuxguitar.app.TGMainSingleton.launchTuxGuitar(TGMainSingleton.java:57)
        at org.herac.tuxguitar.app.TGMainSingleton.launchSingleton(TGMainSingleton.java:45)
        at org.herac.tuxguitar.app.TGMainSingleton.main(TGMainSingleton.java:35)
Caused by: java.util.NoSuchElementException
        at org.herac.tuxguitar.util.TGServiceReader$IteratorImpl.next(TGServiceReader.java:98)
        at org.herac.tuxguitar.util.plugin.TGPluginManager.lookupPlugins(TGPluginManager.java:46)
        ... 12 more
Exception in thread "Thread-1" java.lang.NullPointerException: Cannot invoke "org.herac.tuxguitar.action.TGActionContext.setAttribute(String, Object)" because "actionContext" is null
        at org.herac.tuxguitar.editor.action.TGActionProcessor.appendAttributes(TGActionProcessor.java:67)
        at org.herac.tuxguitar.editor.action.TGActionProcessor.fillActionContext(TGActionProcessor.java:57)
        at org.herac.tuxguitar.editor.action.TGActionProcessor.processOnCurrentThread(TGActionProcessor.java:75)
        at org.herac.tuxguitar.editor.action.TGActionProcessor.processOnCurrentThread(TGActionProcessor.java:88)
        at org.herac.tuxguitar.editor.action.TGActionProcessor$1.run(TGActionProcessor.java:98)
        at org.herac.tuxguitar.thread.TGPooledThreadHandler.processNext(TGPooledThreadHandler.java:66)
        at org.herac.tuxguitar.thread.TGPooledThreadHandler.run(TGPooledThreadHandler.java:86)
        at java.base/java.lang.Thread.run(Thread.java:1583)
`

**Comment by guiv42** [comment]
OK, I've gone too fast :(
Another attempt: https://github.com/guiv42/tuxguitar/commit/38161cd56d6691dc97b792130d7c37a492c0a820

**Comment by G-eos** [comment]
Works well :) except when launched from CLI.
- Double click on a file on the finder -> open it correctly (no empty files). Then double click on another one open the 2nd one :)
- Double click on the app on the finder launch it. Then double click on a file open it (and remove the empty one)
- lance from CLI directly tuxguitar.sh, start with the splash screen. Then double click does not open the file. It seems to launch a 2nd tuxguitar and then die. The first one remain there with no files open in it.

**Comment by guiv42** [comment]
> Works well :) except when launched from CLI.

Very good news, thanks!
This was the targeted behavior. It seems that macOS & SWT do not make the link between the 2 TuxGuitar instances when first is launched from CLI and second from Finder. In this case, swt event listener *is* registered, but it just doesn't receive events. That's really minor from my point of view, most of users just use the graphical interface of the OS.

So, next steps are: cleanup, test again, and merge into master :)

**Comment by G-eos** [comment]
Totally agree. I remove the LockThread in the startup and launch 2 TuxGuitars one from CLI and one from finder. When I double click on a file it opens the one from the finder. And cannot open a 3rd one from finder ;-)

Let me know when you need to test again after clean-up.
Good job !

**Comment by guiv42** [comment]
@G-eos : could you please perform just another test?
Start TuxGuitar from CLI with an argument (path to a valid tg file): does it open correctly? I think this is the only part of opendoc plugin currently untested.

**Comment by G-eos** [comment]
It needed a test ... and some work 

![image](https://github.com/helge17/tuxguitar/assets/1248414/51f60095-cd19-4db1-9d74-89d9d4702c19)


```
2024-04-07 16:31:55.763 java[39511:1928504] NSSoftLinking - The function 'SLSTransactionSetWindowTags' can't be found in the (null) framework.
2024-04-07 16:31:55.763 java[39511:1928504] NSSoftLinking - The function 'SLSTransactionEnsureSpaceSwitchToActiveProcess' can't be found in the (null) framework.
2024-04-07 16:31:55.766 java[39511:1928504] NSSoftLinking - The function 'SLSTransactionReorderWindows' can't be found in the (null) framework.
2024-04-07 16:31:55.767 java[39511:1928504] NSSoftLinking - The function 'SLSTransactionOrderWindowGroupFrontConditionally' can't be found in the (null) framework.
java.lang.NullPointerException: Cannot invoke "org.eclipse.swt.widgets.Display.removeListener(int, org.eclipse.swt.widgets.Listener)" because the return value of "org.eclipse.swt.widgets.Display.getCurrent()" is null
        at org.herac.tuxguitar.cocoa.opendoc.OpenDocListener.disconnect(OpenDocListener.java:55)
        at org.herac.tuxguitar.cocoa.opendoc.OpenDocListener.intercept(OpenDocListener.java:76)
        at org.herac.tuxguitar.action.TGActionManager.intercept(TGActionManager.java:85)
        at org.herac.tuxguitar.action.TGActionManager.execute(TGActionManager.java:64)
        at org.herac.tuxguitar.editor.action.TGActionProcessor.processOnCurrentThread(TGActionProcessor.java:75)
        at org.herac.tuxguitar.editor.action.TGActionProcessor.processOnCurrentThread(TGActionProcessor.java:88)
        at org.herac.tuxguitar.editor.action.TGActionProcessor$1.run(TGActionProcessor.java:98)
        at org.herac.tuxguitar.thread.TGPooledThreadHandler.processNext(TGPooledThreadHandler.java:66)
        at org.herac.tuxguitar.thread.TGPooledThreadHandler.run(TGPooledThreadHandler.java:86)
        at java.base/java.lang.Thread.run(Thread.java:1583)
2024-04-07 16:31:58.701 java[39511:1928504] NSSoftLinking - The function 'SLSTransactionRemoveWindowFromWindowOrderingGroup' can't be found in the (null) framework.
2024-04-07 16:31:59.127 java[39511:1928504] NSSoftLinking - The function 'SLSTransactionAddWindowToWindowOrderingGroup' can't be found in the (null) framework.
2024-04-07 16:32:05.258 java[39511:1928504] NSSoftLinking - The function 'SLSTransactionReorderWindows' can't be found in the (null) framework.
```

**Comment by guiv42** [comment]
OK thanks. Easy to fix.

**Comment by guiv42** [comment]
another test? https://github.com/guiv42/tuxguitar/commit/ae54b3bd000a1b370336e5ca500b4a6311c7c86d

test scenario: call tuxguitar.sh from CLI with an argument (path to a valid tg file)
expected: file opens normally AND no backtrace displayed in CLI

**Comment by guiv42** [comment]
OK, thanks.
Understood: removeListener shall be called from ui thread.

**Comment by guiv42** [comment]
another try
https://github.com/guiv42/tuxguitar/commit/a455b49fb158abbb4ea8af6ef7e33ac1f610a506

**Comment by G-eos** [comment]
Still not good ...

```
2024-04-07 19:41:19.634 java[1366:16737] NSSoftLinking - The function 'SLSTransactionSetWindowTags' can't be found in the (null) framework.
2024-04-07 19:41:19.634 java[1366:16737] NSSoftLinking - The function 'SLSTransactionEnsureSpaceSwitchToActiveProcess' can't be found in the (null) framework.
2024-04-07 19:41:19.638 java[1366:16737] NSSoftLinking - The function 'SLSTransactionReorderWindows' can't be found in the (null) framework.
2024-04-07 19:41:19.638 java[1366:16737] NSSoftLinking - The function 'SLSTransactionOrderWindowGroupFrontConditionally' can't be found in the (null) framework.
java.util.ConcurrentModificationException
        at java.base/java.util.ArrayList$Itr.checkForComodification(ArrayList.java:1095)
        at java.base/java.util.ArrayList$Itr.next(ArrayList.java:1049)
        at org.herac.tuxguitar.action.TGActionManager.intercept(TGActionManager.java:84)
        at org.herac.tuxguitar.action.TGActionManager.execute(TGActionManager.java:64)
        at org.herac.tuxguitar.editor.action.TGActionProcessor.processOnCurrentThread(TGActionProcessor.java:75)
        at org.herac.tuxguitar.editor.action.TGActionProcessor.processOnCurrentThread(TGActionProcessor.java:88)
        at org.herac.tuxguitar.editor.action.TGActionProcessor$1.run(TGActionProcessor.java:98)
        at org.herac.tuxguitar.thread.TGPooledThreadHandler.processNext(TGPooledThreadHandler.java:66)
        at org.herac.tuxguitar.thread.TGPooledThreadHandler.run(TGPooledThreadHandler.java:86)
        at java.base/java.lang.Thread.run(Thread.java:1583)
2024-04-07 19:41:22.595 java[1366:16737] NSSoftLinking - The function 'SLSTransactionRemoveWindowFromWindowOrderingGroup' can't be found in the (null) framework.
```

**Comment by guiv42** [comment]
Okay... so, `Invalid thread access` is solved, but `ConcurrentModificationException` is still there :(

trying to simplify (again), only disconnect plugin when app is closed: https://github.com/guiv42/tuxguitar/commit/9505f051df8f8cee6a52228ae9541e2f27181f89

2 tests to perform here:
- start tuxguitar.sh in CLI with a valid argument -> file is opened, no backtrace printed in CLI (this time I really hope it will be OK)
- close app -> still no backtrace visible in CLI (not so obvious to me...)

**Comment by G-eos** [comment]
it does not want to be kind ...


```
2024-04-07 22:31:34.556 java[2506:47955] NSSoftLinking - The function 'SLSTransactionSetWindowTags' can't be found in the (null) framework.
2024-04-07 22:31:34.557 java[2506:47955] NSSoftLinking - The function 'SLSTransactionEnsureSpaceSwitchToActiveProcess' can't be found in the (null) framework.
2024-04-07 22:31:34.557 java[2506:47955] NSSoftLinking - The function 'SLSTransactionReorderWindows' can't be found in the (null) framework.
2024-04-07 22:31:34.558 java[2506:47955] NSSoftLinking - The function 'SLSTransactionOrderWindowGroupFrontConditionally' can't be found in the (null) framework.
java.util.ConcurrentModificationException
        at java.base/java.util.ArrayList$Itr.checkForComodification(ArrayList.java:1095)
        at java.base/java.util.ArrayList$Itr.next(ArrayList.java:1049)
        at org.herac.tuxguitar.action.TGActionManager.intercept(TGActionManager.java:84)
        at org.herac.tuxguitar.action.TGActionManager.execute(TGActionManager.java:64)
        at org.herac.tuxguitar.editor.action.TGActionProcessor.processOnCurrentThread(TGActionProcessor.java:75)
        at org.herac.tuxguitar.editor.action.TGActionProcessor.processOnCurrentThread(TGActionProcessor.java:88)
        at org.herac.tuxguitar.editor.action.TGActionProcessor$1.run(TGActionProcessor.java:98)
        at org.herac.tuxguitar.thread.TGPooledThreadHandler.processNext(TGPooledThreadHandler.java:66)
        at org.herac.tuxguitar.thread.TGPooledThreadHandler.run(TGPooledThreadHandler.java:86)
        at java.base/java.lang.Thread.run(Thread.java:1583)
2024-04-07 22:31:37.372 java[2506:47955] NSSoftLinking - The function 'SLSTransactionRemoveWindowFromWindowOrderingGroup' can't be found in the (null) framework.
```

**Comment by guiv42** [comment]
:(
but this time I have good news: since it's not related to swt events, I can test and reproduce in my Linux environment. This should be much more efficient!

**Comment by guiv42** [comment]
understood the `ConcurrentModificationException`: cannot remove an interceptor currently processing an action...
https://github.com/guiv42/tuxguitar/commit/7c89434c37f06c2c154b591d36f6b05ed4e3c379

**Comment by G-eos** [comment]
Now its ok :)

- start tuxguitar.sh in CLI with a valid argument -> file is opened, no backtrace printed in CLI
- close app -> still no backtrace visible in CLI 

Works well when double click (when not started on CLI).

Thus, only limitation we have is when it is started through CLI.
Note: why not letting the event registered even if it does not work ? It would be complex later to remember that as it is launched from CLI with arguments the event is no more registered ?

**Comment by guiv42** [comment]
> Note: why not letting the event registered even if it does not work ?

There are 2 different points here:
- the swt event listener (was causing the `Invalid thread access`): this one could remain registered with little or no consequence
- the "action interceptor" (was causing the `ConcurrentModificationException`): this one needs to be removed, or it may catch a following action to create song from default template

And anyway, it's cleaner if both are correctly unregistered when the plugin is disconnected by plugin manager. And here there is no guarantee it will be done from the ui thread (this explains why I wasn't sure about the case when closing app).

Finally it's quite simple, the only complex part here was I wasn't able to test this code by myself. That's why it needed so many iterations. Thanks for the active support!

**Comment by guiv42** [comment]
After some cleanup, a candidate for a pull request (new branch): https://github.com/guiv42/tuxguitar/commit/ab975a4996d213af1cfd901a3380fa5e3816e124
@G-eos : whenever you can test that... I just hope I did not go too fast (it's always a bit risky when I cannot test)

**Comment by G-eos** [comment]
Hi. Works perfectly well.
- From command line: launch tux + splash + open the files, further double click does not open files as usual
- From finder, double click: open a file. further double click open them.
- From finder, launch tuxguitar : ok + further double click open files ...
Note: we can open the same files more than one times ... I do not know if this is a good idea. Tell me if you want a new issue to track it.

**Comment by guiv42** [comment]
OK, thanks -> I will merge this branch into master, and close this issue :)
Since the fix only applies to SWT version, I will create a new issue for JFX version.

>Note: we can open the same files more than one times ... I do not know if this is a good idea. Tell me if you want a new issue to track it.

Yes I agree it deserves a dedicated issue. Note: I can reproduce in Linux configuration.
